persist/fs/write.go

```diff
@@ -195,33 +235,141 @@ func (w *writer) writeAll(
+	}
+
+	// Write the index entries and calculate the bloom filter
+	n, p := uint(w.currIdx), w.bloomFilterFalsePositivePercent
+	m, k := xsets.BloomFilterEstimate(n, p)
+	bloomFilter := xsets.NewBloomFilter(m, k)
+
+	var offset int64
+	for i := range w.indexEntries {
+		id := w.indexEntries[i].id.Data().Get()
+
+		entry := schema.IndexEntry{
+			Index:    w.indexEntries[i].index,
+			ID:       id,
+			Size:     w.indexEntries[i].size,
+			Offset:   w.indexEntries[i].offset,
+			Checksum: w.indexEntries[i].checksum,
+		}
+
+		w.encoder.Reset()
+		if err := w.encoder.EncodeIndexEntry(entry); err != nil {
+			return err
+		}
+
+		data := w.encoder.Bytes()
+		if _, err := w.indexFdWithDigest.Write(data); err != nil {
+			return err
+		}
+
+		// Add to the bloom filter, note this must be zero alloc or else this will
+		// cause heavy GC churn as we flush millions of series at end of each
+		// time window
+		bloomFilter.Add(id)
+
+		if i%summaryEvery == 0 {
+			// Capture the offset for when we write this summary back, only capture
+			// for every summary we'll actually write to avoid a few memcopies
+			w.indexEntries[i].offsetForSummary = offset
+		}
+
+		offset += int64(len(data))
+	}
+
+	// Write summaries and start zeroing out memory to avoid holding onto refs
+	summaries := 0
+	for i := range w.indexEntries {
+		if i%summaryEvery != 0 {
+			w.indexEntries[i].releaseRefs()
+			continue
+		}
+
+		summary := schema.IndexSummary{
+			Index:            w.indexEntries[i].index,
+			ID:               w.indexEntries[i].id.Data().Get(),
+			IndexEntryOffset: w.indexEntries[i].offsetForSummary,
+		}
+
+		w.encoder.Reset()
+		if err := w.encoder.EncodeIndexSummary(summary); err != nil {
+			return err
+		}
+
+		data := w.encoder.Bytes()
+		if _, err := w.summariesFdWithDigest.Write(data); err != nil {
+			return err
+		}
+
+		w.indexEntries[i].releaseRefs()
+		summaries++
+	}
+
+	// Reset summaries slice to avoid allocs for next shard flush, this avoids
+	// leaking memory as we've released all refs while writing summaries
+	w.indexEntries = w.indexEntries[:0]
+
+	// Write the bloom filter bitset out
+	err := bloomFilter.BitSet().Write(w.bloomFilterFdWithDigest)
+	if err != nil {
+		return err
+	}
+
 	info := schema.IndexInfo{
-		Start:     xtime.ToNanoseconds(w.start),
-		BlockSize: int64(w.blockSize),
-		Entries:   w.currIdx,
+		Start:        xtime.ToNanoseconds(w.start),
+		BlockSize:    int64(w.blockSize),
+		Entries:      w.currIdx,
+		MajorVersion: schema.MajorVersion,
+		Summaries: schema.IndexSummariesInfo{
+			Summaries: int64(summaries),
+		},
+		BloomFilter: schema.IndexBloomFilterInfo{
+			NumElementsM: int64(bloomFilter.M()),
+			NumHashesK:   int64(bloomFilter.K()),
+		},
 	}
 
 	w.encoder.Reset()
 	if err := w.encoder.EncodeIndexInfo(info); err != nil {
 		return err
 	}
 
-	if _, err := w.infoFdWithDigest.WriteBytes(w.encoder.Bytes()); err != nil {
+	_, err = w.infoFdWithDigest.Write(w.encoder.Bytes())
```

Comment:
So the info about the index is stored in a separate file than the index itself?

Reply:
Correct, right now all files are either just repeated records or a single record.  i.e. always a single type in one file.